### ISSUE TITLE
Stable call continuations

### ISSUE THREAD

**Issue by nomeata** [dfinity/motoko#2993]
As outlined in https://www.joachim-breitner.de/blog/789-Zero-downtime_upgrades_of_Internet_Computer_canisters I believe that it’s going to be more and more important for canister developers to support zero-downtime upgrades, and don’t have their upgradeability at the whim of other canisters.

With #2942 Motoko already supports this well for canisters that stick to the plain actor model, and only use call one-way methods.

In this issue I’d like to discuss how we can support the full thing: Supporting both instantaneous, non-stopping and safe upgrades _and_ handling responses.

Here is a sketch, based on an idea by @crusso:

## Source

We introduce named callback functions. This is a new visibility, like `private` and `public`. They must be top-level functions of the actor (like methods):
```
actor {
  callback func on_notify(resp : T, more_data: S) : async … {
     …
  }
}
```

Now, an upgradeable call must be of the following form:
```
actor {
      …
      public func do_stuff() : async T {
        let some_stuff_to_remember_about_this_call = …;
        on_notify (await other_canister.notify(args), some_stuff_to_remember_about_this_call) 
      }
}
```
The rules are:

 * All `await` expressions must be immediate arguments to a call to a callback function which is itself in tail-position.
 * All parameter types of a callback function must be “stableable”.

Likely, we want a way to annotate the actor as “should be upgradable” which would make the compiler reject the code if these rules are broken.

## Upgrades

The upgrade signature of a canister must now also carry the information

 * Set of `callback` methods
 * Types of their arguments
 * Types of their results

and the tooling should check that none are dropped, and that their types only change in compatible ways.

## Implementation sketch

The RTS would maintain a map from callback env (as passed to the system and received back upon the callback) to callback name, position of await-ed argument, and other parameters (either given locally, or already received if a call has multiple `await` arguments). This map would be preserved across upgrades in stable memory by the RTS.

While there is a [System API change proposal](https://forum.dfinity.org/t/prioritize-safe-instantaneous-canister-upgrades/9039) (named callbacks) that would make the implementation a bit more natural, we don’t have to wait for this. Because we have full control over the generated wasm, we can keep the callback table index stable, and later use the new System API nicely.

This would give Motoko a Unique Value Proposition that the Rust CDK currently cannot deliver!

**Comment by crusso** [comment]
Thanks for writing this up.

Why not simplify and just use shared functions with shared types? 

Is there any reason to put them in the stable sig? Doesn't candid compatibility give you what you want?

I'd be a little worried about extening to stable types as each message would only preserve internal sharing, not sharing with data-structures in the actor environment, which might lead to confusion.

**Comment by nomeata** [comment]
> Why not simplify and just use shared functions with shared types?

You mean the caniser’s public methods? 

I think you’ll want to distinguish external, untrusted function calls from these continuations, where you have the guarantee that they come as a response to a call you made, and where you can trust the values you pass through.

> Message intended as callbacks could just fail if the sender isn't self.

Callbacks don’t really have a sender on the system level. We could certainly signal “external sender vs. callback” somehow via the context, but I wonder if it isn’t a bit too implicit and too easy to write insecure code. But maybe I am overly worried.

> Is there any reason to put them in the stable sig? Doesn't candid compatibility give you what you want?

Not if we keep the separate from the public entry points. If we don’t do that then … yes, Candid compat might just do the right thing.

> I'd be a little worried about extening to stable types as each message would only preserve internal sharing, not sharing with data-structures in the actor environment, which might lead to confusion.

If the goal is to support calls-across-upgrades, the “environment” has to be stabelizable, I don’t think there is a way around it.
(If we don’t upgrade before the response comes, we wouldn’t serialize of course.)

**Comment by crusso** [comment]
I think this is still too general - in your example, there's some work happening between the await and the call (serializing handling the remaining arguments together with the result of the await). 

I was thinking litterally having  `await f(await e)` as the tail form, in which case both `await e` and `await f(...)` must have shared type. 

I think the example above is probably missing an await... (at least for the current type system).

**Comment by nomeata** [comment]
> some work happening between the await and the call (serializing handling the remaining arguments together with the result of the await).

That's ok, we do get to do some work anyways (e.g. deserializing the replied data and calling the right function) in the RTS's generic callback handler anyway.

And you need to be able to pass data along to the callback handler, else you can't do most of the things you want to be doing.


I probably forgot a tail await.

It may not be fully semantics preserving to essentially compile the tail-await away and let `f` respond (e.g. it may change the reject code). Before `f` had its own call context, afterwards ita execution happens in the current call context. This also changes the cycle balance available, for example.

So callback functions are (slightly) different from public methods.

**Comment by rossberg** [comment]
One big problem I see with this is that it basically makes the await mechanism pointless, since now the programmer has to go back and do their CPS transformation manually after all. So it's too low-level. Can we find a solution that is more declarative, and integrates nicely with the existing await mechanism?

Let me do a little brain dump (which doesn't fully work, see the end).

Part of the issue is to name the underlying callbacks in some deterministic and programmer-controlled fashion, so that they can be mapped to each other across upgrades. We could name awaits themselves for that purpose. Something like:
```
actor {
      …
      public func f() : async T {
        await x1 = a.m1();
        await x2 = b.m2(x1);
        ...
      }
```
Here `await <id> =? <exp>` is a named version of the existing await construct (which, operationally, is a shorthand for `let <id> = await <exp>`). Like other named forms (func, object), it can also be used as an expression:
```
await x2 (b.m2(await x1 (a.m1())))
```
Named awaits can only occur in named functions (or named async expressions, which are extended analogously), and must be unique within them. This creates a deterministic hierarchical namespace for continuations inside an actor from which well-defined callback names could be derived.

The stable signature would declare these in some hierarchical fashion, strawman:
```
actor {
  func f : () -> async T {
    await x1 : Nat;
    await x2 : Text;
  }
}
```
That's a kind of inner protocol for each function. We could even allow nameless awaits, which would create anonymous entries that would match in order (thought that might justify a big fat warning).

The problem is that it is not sufficient: we must also constrain changes to the continuations' closure environments. For that, we would basically need to annotate local variables that are captured in continuation closures. That sounds nasty.

Is there some middle-ground solution perhaps? FWIW, this is closely related to the state-dependency-across-awaits problem. If, for example, we required programmers to be explicit about the variables intended to be used across each await, perhaps that would imply a more natural solution for this problem as well?

**Comment by nomeata** [comment]
> One big problem I see with this is that it basically makes the await mechanism pointless, since now the programmer has to go back and do their CPS transformation manually after all. So it's too low-level.

Yes, absolutely. I am willing to accept that – hot code swapping is likely to have some costs, and I think many of the simpler “smart contracts” are actually better off being written as explicit per-message state machines (easier to reason about, less source of bugs, clear commit semantics).

But that should not stop us from exploring alternatives. The named awaits are an interesting approach, though, and worth exploring. It may be that in the end, with all awaits named and all captured variables given names, it doesn’t add that much any more over manual CPS… but maybe it does.

Also consider `for (i in …) { foo(await x bar()) }`. The CPS translation here turns the loop into a recursive function, and the implicit iterator object becomes part of the captured environment, and thus somehow part of the stable signature. 

Also, addressing the awaits via the sequence of enclosing functions may be too rigid, e.g. when moving code around; or maybe even completely removing calls while still wanting to be able to handle responses.
